core/unix/os.c

```diff
@@ -7585,6 +7601,71 @@ pre_system_call(dcontext_t *dcontext)
 
 #ifdef MACOS
     case SYS_close_nocancel:
+#endif
+#ifdef SYS_close_range
+    case SYS_close_range: {
+        uint first_fd = sys_param(dcontext, 0), last_fd = sys_param(dcontext, 1);
+        uint flags = sys_param(dcontext, 2);
+        bool is_cloexec = TEST(CLOSE_RANGE_CLOEXEC, flags);
+        if (is_cloexec) {
+            /* client.file_io has a test for CLOSE_RANGE_CLOEXEC, but it hasn't been
+             * verified on a system with kernel version >= 5.11 yet.
+             */
+            ASSERT_NOT_TESTED();
+        }
+        /* We do not let the app execute their own close_range ever. Instead we
+         * we make multiple close_range syscalls ourselves, one for each contiguous
+         * range of non-DR-private fds in [first, last].
+         */
+        execute_syscall = false;
+        if (first_fd > last_fd) {
+            set_failure_return_val(dcontext, EINVAL);
+            DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+            break;
+        }
+        uint cur_range_first_fd, cur_range_last_fd;
+        bool cur_range_valid = false;
+        int ret = 0;
+        for (int i = first_fd; i <= last_fd; i++) {
+            /* Do not allow any changes to DR-owned FDs. */
+            if ((is_cloexec && fd_is_dr_owned(i)) ||
```

Comment:
If DR fds are all in a single range, it is more performant to let the close_range syscall execute for the portion that does not overlap.  After all, they created this new syscall purely for performance, so maybe we shouldn't undo it.

For the default -steal_fds DR fds are in a single range -- unless we run out I guess: so even w/ default ops there can be DR fds sprinkled about, although it's unlikely.

I would vote for checking for zero overlap, which should be the common case (probably 90% of cases if people use the rlimit as the upper bound) and letting the close_range syscall execute in that case at least.  That's the fastpath; any overlap at all and we go to this slowpath.

Update: I was misled by the PR description which made me think DR turned the app's close_range into a loop of close calls.  I see this is making close_range calls.  So the only action item then is to maybe tweak the description to make this more clear.

Reply:
The description mentions that we use "multiple such calls, one for each contiguous sub-range". I guess I just need to add "close_range" explicitly there.

Reply:
Done.